{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "143c1466",
   "metadata": {},
   "source": [
    "# Tutorial for performing eQTL analysis for a dataset with many individuals.\n",
    "\n",
    "Most of this tutorial can also be used for testing any feature that is at the replicate/individual level. For example, comparing case vs control would use similar procedure, since the independent variable is defined for each person and not for each cell.\n",
    "\n",
    "To install `memento` in the pre-release version (for Ye Lab members), install it directly from github by running:\n",
    "\n",
    "```pip install git+https://github.com/yelabucsf/scrna-parameter-estimation.git@release-v0.0.8```\n",
    "\n",
    "This requires that you have access to the Ye Lab organization. Be sure to delete any old versions of `memento` via `pip uninstall memento` if you played with the method before."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3d679dbc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import scipy.stats as stats\n",
    "import itertools\n",
    "from pybedtools import BedTool\n",
    "import statsmodels.formula.api as smf\n",
    "import statsmodels.api as sm\n",
    "import imp\n",
    "\n",
    "import os\n",
    "import pickle as pkl\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e325db96",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "# sys.path.append('/home/ssm-user/Github/scrna-parameter-estimation/dist/memento-0.0.8-py3.8.egg')\n",
    "import memento"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1b932a18",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path  = '/data_volume/memento/lupus/'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "efe40fba",
   "metadata": {},
   "source": [
    "### Read the inputs: variables of interest (SNPs), covariates, SNP-gene pairs.\n",
    "\n",
    "For each of these SNP and covariate, each row is an individual and columns are different variables of interest. \n",
    "\n",
    "For the tutorial, we use the genotypes and covariates used in 2022 Perez, Gordon, Subramaniam et al. paper from the lab. These inputs are identical to Matrix eQTL inputs - I just transpose them here because I think it makes more sense that observations are rows...\n",
    "\n",
    "For the tutorial, we just setup some random SNP-gene pairs to test; however, you can flexibly design this mapping to fit your needs. I purposefully didn't encode all possible variations of how you can define gene-SNP relationships."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "120d0011",
   "metadata": {},
   "outputs": [],
   "source": [
    "pop = 'eur'\n",
    "snps_path = data_path + 'mateqtl_input/{}_genos.tsv'.format(pop)\n",
    "cov_path = data_path + 'mateqtl_input/{}_mateqtl_cov.txt'.format(pop)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e78abf7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "snps = pd.read_csv(snps_path, sep='\\t', index_col=0).T\n",
    "cov = pd.read_csv(cov_path, sep='\\t', index_col=0).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9797dadb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>CHROM:POS</th>\n",
       "      <th>3:165182446</th>\n",
       "      <th>6:122682327</th>\n",
       "      <th>22:40561759</th>\n",
       "      <th>3:104381193</th>\n",
       "      <th>15:57107863</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1132_1132</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285_1285</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1961_1961</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HC-526</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1414_1414</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "CHROM:POS  3:165182446  6:122682327  22:40561759  3:104381193  15:57107863\n",
       "1132_1132            1            1            1            1            2\n",
       "1285_1285            2            1            2            1            2\n",
       "1961_1961            1            2            2            0            2\n",
       "HC-526               1            1            2            2            1\n",
       "1414_1414            0            0            2            1            2"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Print the first 5 SNPs for the first 5 individuals to show the structure\n",
    "snps.iloc[:, :5].head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "fc431649",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>age</th>\n",
       "      <th>Female</th>\n",
       "      <th>status</th>\n",
       "      <th>PC1_e</th>\n",
       "      <th>PC2_e</th>\n",
       "      <th>PC3_e</th>\n",
       "      <th>PC4_e</th>\n",
       "      <th>PC5_e</th>\n",
       "      <th>PC6_e</th>\n",
       "      <th>PC7_e</th>\n",
       "      <th>...</th>\n",
       "      <th>batch_cov_b_14</th>\n",
       "      <th>batch_cov_b_15</th>\n",
       "      <th>batch_cov_b_2</th>\n",
       "      <th>batch_cov_b_3</th>\n",
       "      <th>batch_cov_b_4</th>\n",
       "      <th>batch_cov_b_5</th>\n",
       "      <th>batch_cov_b_6</th>\n",
       "      <th>batch_cov_b_7</th>\n",
       "      <th>batch_cov_b_8</th>\n",
       "      <th>batch_cov_b_9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1132_1132</th>\n",
       "      <td>45.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>19.067178</td>\n",
       "      <td>17.787198</td>\n",
       "      <td>10.275343</td>\n",
       "      <td>-2.829570</td>\n",
       "      <td>-3.546597</td>\n",
       "      <td>-1.269196</td>\n",
       "      <td>-2.183796</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285_1285</th>\n",
       "      <td>39.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14.471841</td>\n",
       "      <td>18.737343</td>\n",
       "      <td>12.465061</td>\n",
       "      <td>11.195105</td>\n",
       "      <td>-2.246129</td>\n",
       "      <td>-11.168822</td>\n",
       "      <td>2.230269</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1961_1961</th>\n",
       "      <td>43.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-7.343628</td>\n",
       "      <td>38.241007</td>\n",
       "      <td>-7.431836</td>\n",
       "      <td>0.607220</td>\n",
       "      <td>-13.730105</td>\n",
       "      <td>-2.339229</td>\n",
       "      <td>-3.238375</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HC-526</th>\n",
       "      <td>58.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.495487</td>\n",
       "      <td>-17.795535</td>\n",
       "      <td>0.458286</td>\n",
       "      <td>5.384761</td>\n",
       "      <td>-10.269823</td>\n",
       "      <td>-2.239953</td>\n",
       "      <td>-4.240055</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1414_1414</th>\n",
       "      <td>41.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-10.313840</td>\n",
       "      <td>-3.423322</td>\n",
       "      <td>1.635042</td>\n",
       "      <td>9.192646</td>\n",
       "      <td>-3.507571</td>\n",
       "      <td>11.446228</td>\n",
       "      <td>-3.834848</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            age  Female  status      PC1_e      PC2_e      PC3_e      PC4_e  \\\n",
       "1132_1132  45.0     0.0     0.0  19.067178  17.787198  10.275343  -2.829570   \n",
       "1285_1285  39.0     0.0     0.0  14.471841  18.737343  12.465061  11.195105   \n",
       "1961_1961  43.0     0.0     0.0  -7.343628  38.241007  -7.431836   0.607220   \n",
       "HC-526     58.0     0.0     1.0   0.495487 -17.795535   0.458286   5.384761   \n",
       "1414_1414  41.0     1.0     0.0 -10.313840  -3.423322   1.635042   9.192646   \n",
       "\n",
       "               PC5_e      PC6_e     PC7_e  ...  batch_cov_b_14  \\\n",
       "1132_1132  -3.546597  -1.269196 -2.183796  ...             0.0   \n",
       "1285_1285  -2.246129 -11.168822  2.230269  ...             1.0   \n",
       "1961_1961 -13.730105  -2.339229 -3.238375  ...             0.0   \n",
       "HC-526    -10.269823  -2.239953 -4.240055  ...             0.0   \n",
       "1414_1414  -3.507571  11.446228 -3.834848  ...             0.0   \n",
       "\n",
       "           batch_cov_b_15  batch_cov_b_2  batch_cov_b_3  batch_cov_b_4  \\\n",
       "1132_1132             1.0            0.0            0.0            0.0   \n",
       "1285_1285             0.0            0.0            0.0            0.0   \n",
       "1961_1961             0.0            0.0            0.0            0.0   \n",
       "HC-526                0.0            0.0            0.0            0.0   \n",
       "1414_1414             0.0            0.0            0.0            0.0   \n",
       "\n",
       "           batch_cov_b_5  batch_cov_b_6  batch_cov_b_7  batch_cov_b_8  \\\n",
       "1132_1132            0.0            0.0            0.0            0.0   \n",
       "1285_1285            0.0            0.0            0.0            0.0   \n",
       "1961_1961            0.0            0.0            1.0            0.0   \n",
       "HC-526               0.0            0.0            0.0            0.0   \n",
       "1414_1414            0.0            1.0            0.0            0.0   \n",
       "\n",
       "           batch_cov_b_9  \n",
       "1132_1132            0.0  \n",
       "1285_1285            0.0  \n",
       "1961_1961            0.0  \n",
       "HC-526               0.0  \n",
       "1414_1414            0.0  \n",
       "\n",
       "[5 rows x 35 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Print the covariates for the first 5 individuals to show the structure\n",
    "cov.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2f765e15",
   "metadata": {},
   "outputs": [],
   "source": [
    "# You can define this mapping DataFrame however you want - for example, you can take find gene-SNP pairs via looking for a 100kb.\n",
    "# Here, to make the tutorial faster, we'll just randomly take 50k lines.\n",
    "gene_snp_pairs = pd.read_csv(data_path + 'mateqtl_input/{}/gene_snp_pairs_hg19_100kb.csv'.format(pop))\n",
    "gene_snp_pairs.columns = ['gene', 'SNP']\n",
    "gene_snp_pairs = gene_snp_pairs.query('SNP in @snps.columns').sample(50000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "124d534b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>SNP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5458273</th>\n",
       "      <td>PEX3</td>\n",
       "      <td>6:143704295</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4909714</th>\n",
       "      <td>SLC30A5</td>\n",
       "      <td>5:68296445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3722035</th>\n",
       "      <td>BOK</td>\n",
       "      <td>2:242503543</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3540175</th>\n",
       "      <td>AMER3</td>\n",
       "      <td>2:131610074</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3619623</th>\n",
       "      <td>MYO1B</td>\n",
       "      <td>2:192016879</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>816583</th>\n",
       "      <td>VAX1</td>\n",
       "      <td>10:118985613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5928785</th>\n",
       "      <td>RP1</td>\n",
       "      <td>8:55598202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>493305</th>\n",
       "      <td>C1orf116</td>\n",
       "      <td>1:207165730</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2422224</th>\n",
       "      <td>ZMYND15</td>\n",
       "      <td>17:4690500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>583777</th>\n",
       "      <td>OR2T33</td>\n",
       "      <td>1:248481384</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene           SNP\n",
       "5458273      PEX3   6:143704295\n",
       "4909714   SLC30A5    5:68296445\n",
       "3722035       BOK   2:242503543\n",
       "3540175     AMER3   2:131610074\n",
       "3619623     MYO1B   2:192016879\n",
       "816583       VAX1  10:118985613\n",
       "5928785       RP1    8:55598202\n",
       "493305   C1orf116   1:207165730\n",
       "2422224   ZMYND15    17:4690500\n",
       "583777     OR2T33   1:248481384"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_snp_pairs.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99dccd56",
   "metadata": {},
   "source": [
    "### Read h5ad object\n",
    "\n",
    "Standard h5ad file in the scanpy workflow. Some things to keep in mind:\n",
    "\n",
    "- `adata.X` should be the raw counts with all genes detected. Typically, this will be the size of N cells with ~30k genes in a standard 10X experiment. \n",
    "- Here, we will just use the T4 cells defined by one of the AnnData.obs columns.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "db428899",
   "metadata": {},
   "outputs": [],
   "source": [
    "ct = 'T4'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "775d69f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read(data_path + 'single_cell/{}_{}.h5ad'.format(pop, ct))\n",
    "adata = adata[adata.obs.ind_cov.isin(snps.index)].copy() # pick out individuals we have genotype and covariates for\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "99e038ff",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We have 129531 cells labeled as T4\n"
     ]
    }
   ],
   "source": [
    "print('We have {} cells labeled as T4'.format(adata.shape[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "64ed56bb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>batch_cov</th>\n",
       "      <th>ind_cov</th>\n",
       "      <th>Processing_Cohort</th>\n",
       "      <th>louvain</th>\n",
       "      <th>cg_cov</th>\n",
       "      <th>ct_cov</th>\n",
       "      <th>L3</th>\n",
       "      <th>ind_cov_batch_cov</th>\n",
       "      <th>Age</th>\n",
       "      <th>Sex</th>\n",
       "      <th>pop_cov</th>\n",
       "      <th>Status</th>\n",
       "      <th>SLE_status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GTCACGGAGATTACCC-1-1-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-1-0-0-0-0-0-0-0-0-0-0-0-0-0</th>\n",
       "      <td>dmx_YE_8-2</td>\n",
       "      <td>1368_1368</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1</td>\n",
       "      <td>T4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1368_1368:dmx_YE_8-2</td>\n",
       "      <td>45.0</td>\n",
       "      <td>Male</td>\n",
       "      <td>European</td>\n",
       "      <td>Managed</td>\n",
       "      <td>SLE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTCATTTCAGAGTGTG-1-1-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-1-0-0-0-0-0-0</th>\n",
       "      <td>dmx_YS-JY-22_pool5</td>\n",
       "      <td>HC-540</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2</td>\n",
       "      <td>T4</td>\n",
       "      <td>T4_em</td>\n",
       "      <td>1.0</td>\n",
       "      <td>HC-540:dmx_YS-JY-22_pool5</td>\n",
       "      <td>68.0</td>\n",
       "      <td>Female</td>\n",
       "      <td>European</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>Healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAGATGGTTCACGGC-1-1-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-0-1-0-0-0-0-0-0-0-0-0-0</th>\n",
       "      <td>dmx_YS-JY-20_pool3</td>\n",
       "      <td>HC-006</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1</td>\n",
       "      <td>T4</td>\n",
       "      <td>T4_naive</td>\n",
       "      <td>1.0</td>\n",
       "      <td>HC-006:dmx_YS-JY-20_pool3</td>\n",
       "      <td>53.0</td>\n",
       "      <td>Female</td>\n",
       "      <td>European</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>Healthy</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                             batch_cov  \\\n",
       "GTCACGGAGATTACCC-1-1-0-0-0-0-0-0-0-0-0-0-0-0-0-...          dmx_YE_8-2   \n",
       "GTCATTTCAGAGTGTG-1-1-0-0-0-0-0-0-0-0-0-0-0-0-0-...  dmx_YS-JY-22_pool5   \n",
       "AAAGATGGTTCACGGC-1-1-0-0-0-0-0-0-0-0-0-0-0-0-0-...  dmx_YS-JY-20_pool3   \n",
       "\n",
       "                                                      ind_cov  \\\n",
       "GTCACGGAGATTACCC-1-1-0-0-0-0-0-0-0-0-0-0-0-0-0-...  1368_1368   \n",
       "GTCATTTCAGAGTGTG-1-1-0-0-0-0-0-0-0-0-0-0-0-0-0-...     HC-540   \n",
       "AAAGATGGTTCACGGC-1-1-0-0-0-0-0-0-0-0-0-0-0-0-0-...     HC-006   \n",
       "\n",
       "                                                   Processing_Cohort louvain  \\\n",
       "GTCACGGAGATTACCC-1-1-0-0-0-0-0-0-0-0-0-0-0-0-0-...               2.0       1   \n",
       "GTCATTTCAGAGTGTG-1-1-0-0-0-0-0-0-0-0-0-0-0-0-0-...               4.0       2   \n",
       "AAAGATGGTTCACGGC-1-1-0-0-0-0-0-0-0-0-0-0-0-0-0-...               4.0       1   \n",
       "\n",
       "                                                   cg_cov    ct_cov   L3  \\\n",
       "GTCACGGAGATTACCC-1-1-0-0-0-0-0-0-0-0-0-0-0-0-0-...     T4       NaN  0.0   \n",
       "GTCATTTCAGAGTGTG-1-1-0-0-0-0-0-0-0-0-0-0-0-0-0-...     T4     T4_em  1.0   \n",
       "AAAGATGGTTCACGGC-1-1-0-0-0-0-0-0-0-0-0-0-0-0-0-...     T4  T4_naive  1.0   \n",
       "\n",
       "                                                            ind_cov_batch_cov  \\\n",
       "GTCACGGAGATTACCC-1-1-0-0-0-0-0-0-0-0-0-0-0-0-0-...       1368_1368:dmx_YE_8-2   \n",
       "GTCATTTCAGAGTGTG-1-1-0-0-0-0-0-0-0-0-0-0-0-0-0-...  HC-540:dmx_YS-JY-22_pool5   \n",
       "AAAGATGGTTCACGGC-1-1-0-0-0-0-0-0-0-0-0-0-0-0-0-...  HC-006:dmx_YS-JY-20_pool3   \n",
       "\n",
       "                                                     Age     Sex   pop_cov  \\\n",
       "GTCACGGAGATTACCC-1-1-0-0-0-0-0-0-0-0-0-0-0-0-0-...  45.0    Male  European   \n",
       "GTCATTTCAGAGTGTG-1-1-0-0-0-0-0-0-0-0-0-0-0-0-0-...  68.0  Female  European   \n",
       "AAAGATGGTTCACGGC-1-1-0-0-0-0-0-0-0-0-0-0-0-0-0-...  53.0  Female  European   \n",
       "\n",
       "                                                     Status SLE_status  \n",
       "GTCACGGAGATTACCC-1-1-0-0-0-0-0-0-0-0-0-0-0-0-0-...  Managed        SLE  \n",
       "GTCATTTCAGAGTGTG-1-1-0-0-0-0-0-0-0-0-0-0-0-0-0-...  Healthy    Healthy  \n",
       "AAAGATGGTTCACGGC-1-1-0-0-0-0-0-0-0-0-0-0-0-0-0-...  Healthy    Healthy  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "972cc57e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<5x15 sparse matrix of type '<class 'numpy.float32'>'\n",
       "\twith 0 stored elements in Compressed Sparse Row format>"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.X[:5, :15]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "d14040a8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Confirming that adata.X is a sparse matrix of counts.\n",
      "Row sums are:\n",
      "[[1905.]\n",
      " [2104.]\n",
      " [2102.]\n",
      " [1209.]\n",
      " [2030.]]\n",
      "\n",
      "The matrix itself:\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<129531x32738 sparse matrix of type '<class 'numpy.float32'>'\n",
       "\twith 83322139 stored elements in Compressed Sparse Row format>"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# adata.X should be a sparse matrix with counts\n",
    "print('Confirming that adata.X is a sparse matrix of counts.')\n",
    "print('Row sums are:')\n",
    "print(adata.X.sum(axis=1)[:5])\n",
    "print('')\n",
    "print('The matrix itself:')\n",
    "adata.X"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf5b0d90",
   "metadata": {},
   "source": [
    "### Define the helper function for running `memento` on genetics data\n",
    "\n",
    "Next iteration maybe this will be a function in the package. For now, define the function manually."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "1cefb1ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_genetics(adata, snps, cov, gene_snp_pairs, num_cpu, num_blocks=5, num_boot=5000):\n",
    "    \"\"\"\n",
    "    Run mean eQTL analysis using memento. adata, snp, cov, gene_snp_pairs should be specified in the format above.\n",
    "    \"\"\"\n",
    "    \n",
    "    # Setup memento\n",
    "    adata.obs['capture_rate'] = 0.1\n",
    "    memento.setup_memento(adata, q_column ='capture_rate', trim_percent=0.1, filter_mean_thresh=0.05, estimator_type='mean_only')\n",
    "    memento.create_groups(adata, label_columns=['ind_cov'])\n",
    "    \n",
    "    # To prevent memory issues, we will separate out some SNPs for each gene\n",
    "    reordered_gene_snp_pairs = gene_snp_pairs.sample(frac=1)\n",
    "    \n",
    "    # Re-order the SNP and covariate DataFrames to the order that memento expects\n",
    "    sample_order = memento.get_groups(adata) #the order of samples that memento expects\n",
    "    cov_df = cov.loc[sample_order['ind_cov']]\n",
    "    snps_df = snps.loc[sample_order['ind_cov']]\n",
    "    \n",
    "    # Number of tests in each block\n",
    "    blocksize=int(gene_snp_pairs.shape[0]/num_blocks)\n",
    "    print('blocksize', blocksize)\n",
    "    \n",
    "    # Run memento for each block\n",
    "    results = []\n",
    "    for block in range(num_blocks):\n",
    "        print('working on block', block)\n",
    "        subset_gene_snp_pairs = reordered_gene_snp_pairs.iloc[(blocksize*block):(blocksize*(block+1)), :]\n",
    "\n",
    "        memento.compute_1d_moments(adata, min_perc_group=.5, gene_list=subset_gene_snp_pairs.gene.drop_duplicates().tolist())\n",
    "\n",
    "        gene_to_snp_dictionary = dict(subset_gene_snp_pairs[subset_gene_snp_pairs.gene.isin(adata.var.index)].groupby('gene')['SNP'].apply(list))\n",
    "\n",
    "        memento.ht_1d_moments(\n",
    "            adata, \n",
    "            covariate=cov_df,\n",
    "            treatment=snps_df,\n",
    "            treatment_for_gene=gene_to_snp_dictionary,\n",
    "            num_boot=num_boot,\n",
    "            verbose=1,\n",
    "            num_cpus=num_cpu,\n",
    "            resampling='bootstrap',\n",
    "            approx=True,\n",
    "            resample_rep=True)\n",
    "        results.append(memento.get_1d_ht_result(adata))\n",
    "                                                \n",
    "    return pd.concat(results)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5975a7a2",
   "metadata": {},
   "source": [
    "### Run memento\n",
    "\n",
    "Due to the resampling at the single-cell level, `memento` generally takes much longer than something like Matrix eQTL or FastQTL that works on pseudobulks. It is however much faster than fitting linear mixed models for millions of cells. \n",
    "\n",
    "I would recommend using as many cores as you can afford on a high performance computing cluster. (for members of the Ye lab, maybe something like c5.24xlarge instance just for a few hours for this and set the num_cpu to 80). \n",
    "\n",
    "`memento` is fairly flexible in that if you're willing to use more CPUs, the speed will scale linearly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "44b4a83f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "blocksize 10000\n",
      "working on block 0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Parallel(n_jobs=10)]: Using backend LokyBackend with 10 concurrent workers.\n",
      "[Parallel(n_jobs=10)]: Done  30 tasks      | elapsed:   11.4s\n",
      "[Parallel(n_jobs=10)]: Done 180 tasks      | elapsed:   46.7s\n",
      "[Parallel(n_jobs=10)]: Done 430 tasks      | elapsed:  1.8min\n",
      "[Parallel(n_jobs=10)]: Done 780 tasks      | elapsed:  3.2min\n",
      "[Parallel(n_jobs=10)]: Done 947 out of 947 | elapsed:  3.9min finished\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "working on block 1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Parallel(n_jobs=10)]: Using backend LokyBackend with 10 concurrent workers.\n",
      "[Parallel(n_jobs=10)]: Done  30 tasks      | elapsed:    8.6s\n",
      "[Parallel(n_jobs=10)]: Done 180 tasks      | elapsed:   46.3s\n",
      "[Parallel(n_jobs=10)]: Done 408 out of 408 | elapsed:  1.7min finished\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "working on block 2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Parallel(n_jobs=10)]: Using backend LokyBackend with 10 concurrent workers.\n",
      "[Parallel(n_jobs=10)]: Done  30 tasks      | elapsed:    8.6s\n",
      "[Parallel(n_jobs=10)]: Done 193 out of 193 | elapsed:   48.4s finished\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "working on block 3\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Parallel(n_jobs=10)]: Using backend LokyBackend with 10 concurrent workers.\n",
      "[Parallel(n_jobs=10)]: Done  30 tasks      | elapsed:    8.8s\n",
      "[Parallel(n_jobs=10)]: Done  97 out of  97 | elapsed:   26.3s finished\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "working on block 4\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Parallel(n_jobs=10)]: Using backend LokyBackend with 10 concurrent workers.\n",
      "[Parallel(n_jobs=10)]: Done  30 tasks      | elapsed:    9.3s\n",
      "[Parallel(n_jobs=10)]: Done  54 out of  54 | elapsed:   17.4s finished\n"
     ]
    }
   ],
   "source": [
    "eqtl_results = run_genetics(\n",
    "    adata=adata,\n",
    "    snps=snps,\n",
    "    cov=cov,\n",
    "    gene_snp_pairs=gene_snp_pairs,\n",
    "    num_cpu=10,\n",
    "    num_blocks=5, # increase this if you run out of memory.\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "c865df2e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>tx</th>\n",
       "      <th>de_coef</th>\n",
       "      <th>de_se</th>\n",
       "      <th>de_pval</th>\n",
       "      <th>dv_coef</th>\n",
       "      <th>dv_se</th>\n",
       "      <th>dv_pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NOC2L</td>\n",
       "      <td>1:845635</td>\n",
       "      <td>0.005111</td>\n",
       "      <td>0.002424</td>\n",
       "      <td>0.035050</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.967713e-17</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NOC2L</td>\n",
       "      <td>1:881627</td>\n",
       "      <td>-0.005168</td>\n",
       "      <td>0.001797</td>\n",
       "      <td>0.004051</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.361005e-17</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SDF4</td>\n",
       "      <td>1:1062025</td>\n",
       "      <td>0.005272</td>\n",
       "      <td>0.002748</td>\n",
       "      <td>0.055024</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.480994e-17</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>UBE2J2</td>\n",
       "      <td>1:1161641</td>\n",
       "      <td>-0.000393</td>\n",
       "      <td>0.002382</td>\n",
       "      <td>0.869094</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.827615e-17</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CCNL2</td>\n",
       "      <td>1:1252309</td>\n",
       "      <td>-0.004638</td>\n",
       "      <td>0.002607</td>\n",
       "      <td>0.075280</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.492581e-17</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RER1</td>\n",
       "      <td>1:2229156</td>\n",
       "      <td>-0.001183</td>\n",
       "      <td>0.003967</td>\n",
       "      <td>0.765610</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.040996e-17</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TNFRSF14</td>\n",
       "      <td>1:2560903</td>\n",
       "      <td>-0.003125</td>\n",
       "      <td>0.003016</td>\n",
       "      <td>0.300386</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.688472e-17</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RPL22</td>\n",
       "      <td>1:6317420</td>\n",
       "      <td>0.005040</td>\n",
       "      <td>0.013121</td>\n",
       "      <td>0.701416</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.006726e-17</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>1:7895069</td>\n",
       "      <td>-0.000752</td>\n",
       "      <td>0.002356</td>\n",
       "      <td>0.750012</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.653482e-17</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>1:6846552</td>\n",
       "      <td>-0.003309</td>\n",
       "      <td>0.002685</td>\n",
       "      <td>0.217853</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.304167e-17</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene         tx   de_coef     de_se   de_pval  dv_coef         dv_se  \\\n",
       "0     NOC2L   1:845635  0.005111  0.002424  0.035050      0.0  5.967713e-17   \n",
       "1     NOC2L   1:881627 -0.005168  0.001797  0.004051      0.0  4.361005e-17   \n",
       "0      SDF4  1:1062025  0.005272  0.002748  0.055024      0.0  6.480994e-17   \n",
       "0    UBE2J2  1:1161641 -0.000393  0.002382  0.869094      0.0  5.827615e-17   \n",
       "0     CCNL2  1:1252309 -0.004638  0.002607  0.075280      0.0  6.492581e-17   \n",
       "0      RER1  1:2229156 -0.001183  0.003967  0.765610      0.0  9.040996e-17   \n",
       "0  TNFRSF14  1:2560903 -0.003125  0.003016  0.300386      0.0  6.688472e-17   \n",
       "0     RPL22  1:6317420  0.005040  0.013121  0.701416      0.0  6.006726e-17   \n",
       "0    CAMTA1  1:7895069 -0.000752  0.002356  0.750012      0.0  5.653482e-17   \n",
       "1    CAMTA1  1:6846552 -0.003309  0.002685  0.217853      0.0  6.304167e-17   \n",
       "\n",
       "   dv_pval  \n",
       "0      1.0  \n",
       "1      1.0  \n",
       "0      1.0  \n",
       "0      1.0  \n",
       "0      1.0  \n",
       "0      1.0  \n",
       "0      1.0  \n",
       "0      1.0  \n",
       "0      1.0  \n",
       "1      1.0  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# The de_pval is not corrected on purpose - user can correct the P-values however they please.\n",
    "eqtl_results.head(10)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
