{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Including covariates for differential gene expression testing using `memento`\n",
    "\n",
    "To install `memento` in the pre-release version (for Ye Lab members), install it directly from github by running:\n",
    "\n",
    "```pip install git+https://github.com/yelabucsf/scrna-parameter-estimation.git@release-v0.0.8```\n",
    "\n",
    "This requires that you have access to the Ye Lab organization. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This is only for development purposes\n",
    "\n",
    "import sys\n",
    "# sys.path.append('/home/ssm-user/Github/scrna-parameter-estimation/dist/memento-0.0.8-py3.8.egg')\n",
    "import memento\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import memento"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig_path = '~/Github/scrna-parameter-estimation/figures/fig4/'\n",
    "data_path = '/data_volume/memento/demux/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle as pkl"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read IFN data and filter for monocytes\n",
    "\n",
    "For `memento`, we need the raw count matrix. Preferrably, feed the one with all genes so that we can choose what genes to look at. \n",
    "\n",
    "One of the columns in `adata.obs` should be the discrete groups to compare mean, variability, and co-variability across. In this case, it's called `stim`. \n",
    "\n",
    "The column containing the covariate that you want p-values for should either:\n",
    "- Be binary (aka the column only contains two unique values, such as 'A' and 'B'. Here, the values are either 'stim' or 'ctrl'.\n",
    "- Be numeric (aka the column contains -1, 0, -1 for each genotype value). \n",
    "\n",
    "I recommend changing the labels to something numeric (here, i use 0 for `ctrl` and 1 for `stim`). Otherwise, the sign of the DE/EV/DC testing will be very hard to interpret."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ssm-user/anaconda3/envs/single_cell/lib/python3.8/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n"
     ]
    }
   ],
   "source": [
    "adata = sc.read(data_path + 'interferon_filtered.h5ad')\n",
    "adata = adata[adata.obs.cell == 'CD14+ Monocytes'].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['stim'] = adata.obs['stim'].apply(lambda x: 0 if x == 'ctrl' else 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ind</th>\n",
       "      <th>stim</th>\n",
       "      <th>cell</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GACGTATGTAACCG-1</th>\n",
       "      <td>1015</td>\n",
       "      <td>1</td>\n",
       "      <td>CD14+ Monocytes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GACATTCTGGGACA-1</th>\n",
       "      <td>1256</td>\n",
       "      <td>0</td>\n",
       "      <td>CD14+ Monocytes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TACCGAGATCCTGC-1</th>\n",
       "      <td>1488</td>\n",
       "      <td>1</td>\n",
       "      <td>CD14+ Monocytes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGATACCTGCCTTC-1</th>\n",
       "      <td>1015</td>\n",
       "      <td>1</td>\n",
       "      <td>CD14+ Monocytes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTATGTTGTAGACC-1</th>\n",
       "      <td>1016</td>\n",
       "      <td>0</td>\n",
       "      <td>CD14+ Monocytes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   ind stim             cell\n",
       "index                                       \n",
       "GACGTATGTAACCG-1  1015    1  CD14+ Monocytes\n",
       "GACATTCTGGGACA-1  1256    0  CD14+ Monocytes\n",
       "TACCGAGATCCTGC-1  1488    1  CD14+ Monocytes\n",
       "TGATACCTGCCTTC-1  1015    1  CD14+ Monocytes\n",
       "CTATGTTGTAGACC-1  1016    0  CD14+ Monocytes"
      ]
     },
     "execution_count": 168,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs[['ind', 'stim', 'cell']].sample(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Engineer the covariate to be used in memento. \n",
    "\n",
    "Currently, optimizations in `memento` only supports discrete covariates, with fewer covariates the better. Here, we are interested in whether the stimulation affects gene expression on chromosome 1, while also including the total chromosome 1 count as a covariate.\n",
    "\n",
    "There are now 10 unique values for \"chr_expr_avg\" column in `adata.obs`. We will use this as the covariate."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ssm-user/anaconda3/envs/single_cell/lib/python3.8/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n"
     ]
    }
   ],
   "source": [
    "# These are not actually chromosome 1 genes\n",
    "# TODO: Remake this tutorial with actual chr1 labels and maybe with the aneuploidy dataset so that it makes sense\n",
    "chr1_genes = list(np.random.choice(adata.var.index, 4000))\n",
    "\n",
    "adata_chrom = adata.copy().copy()\n",
    "adata_chrom.obs['chr_expr'] = adata_chrom[:, chr1_genes].X.sum(axis=1).astype(int)\n",
    "adata_chrom.obs['chr_expr_bin'] = pd.qcut(adata_chrom.obs['chr_expr'], 10)\n",
    "adata_chrom.obs = adata_chrom.obs.join(adata_chrom.obs.groupby('chr_expr_bin')['chr_expr'].median(), on='chr_expr_bin', rsuffix='_avg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tsne1</th>\n",
       "      <th>tsne2</th>\n",
       "      <th>ind</th>\n",
       "      <th>stim</th>\n",
       "      <th>cluster</th>\n",
       "      <th>cell</th>\n",
       "      <th>multiplets</th>\n",
       "      <th>n_genes_by_counts</th>\n",
       "      <th>log1p_n_genes_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>...</th>\n",
       "      <th>total_counts_mt</th>\n",
       "      <th>log1p_total_counts_mt</th>\n",
       "      <th>pct_counts_mt</th>\n",
       "      <th>total_counts_hb</th>\n",
       "      <th>log1p_total_counts_hb</th>\n",
       "      <th>pct_counts_hb</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>chr_expr</th>\n",
       "      <th>chr_expr_bin</th>\n",
       "      <th>chr_expr_avg</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACATACATTTCC-1</th>\n",
       "      <td>-27.640373</td>\n",
       "      <td>14.966629</td>\n",
       "      <td>1016</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>CD14+ Monocytes</td>\n",
       "      <td>singlet</td>\n",
       "      <td>878</td>\n",
       "      <td>6.778785</td>\n",
       "      <td>3018.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>CD14+ Monocytes - ctrl</td>\n",
       "      <td>365</td>\n",
       "      <td>(325.0, 377.0]</td>\n",
       "      <td>352.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATACCAGAAA-1</th>\n",
       "      <td>-27.493646</td>\n",
       "      <td>28.924885</td>\n",
       "      <td>1256</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>CD14+ Monocytes</td>\n",
       "      <td>singlet</td>\n",
       "      <td>713</td>\n",
       "      <td>6.570883</td>\n",
       "      <td>2481.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>CD14+ Monocytes - ctrl</td>\n",
       "      <td>314</td>\n",
       "      <td>(280.0, 325.0]</td>\n",
       "      <td>304.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      tsne1      tsne2   ind stim cluster             cell  \\\n",
       "index                                                                        \n",
       "AAACATACATTTCC-1 -27.640373  14.966629  1016    0       9  CD14+ Monocytes   \n",
       "AAACATACCAGAAA-1 -27.493646  28.924885  1256    0       9  CD14+ Monocytes   \n",
       "\n",
       "                 multiplets  n_genes_by_counts  log1p_n_genes_by_counts  \\\n",
       "index                                                                     \n",
       "AAACATACATTTCC-1    singlet                878                 6.778785   \n",
       "AAACATACCAGAAA-1    singlet                713                 6.570883   \n",
       "\n",
       "                  total_counts  ...  total_counts_mt  log1p_total_counts_mt  \\\n",
       "index                           ...                                           \n",
       "AAACATACATTTCC-1        3018.0  ...              0.0                    0.0   \n",
       "AAACATACCAGAAA-1        2481.0  ...              0.0                    0.0   \n",
       "\n",
       "                  pct_counts_mt  total_counts_hb  log1p_total_counts_hb  \\\n",
       "index                                                                     \n",
       "AAACATACATTTCC-1            0.0              0.0                    0.0   \n",
       "AAACATACCAGAAA-1            0.0              0.0                    0.0   \n",
       "\n",
       "                  pct_counts_hb               cell_type chr_expr  \\\n",
       "index                                                              \n",
       "AAACATACATTTCC-1            0.0  CD14+ Monocytes - ctrl      365   \n",
       "AAACATACCAGAAA-1            0.0  CD14+ Monocytes - ctrl      314   \n",
       "\n",
       "                    chr_expr_bin chr_expr_avg  \n",
       "index                                          \n",
       "AAACATACATTTCC-1  (325.0, 377.0]        352.0  \n",
       "AAACATACCAGAAA-1  (280.0, 325.0]        304.0  \n",
       "\n",
       "[2 rows x 21 columns]"
      ]
     },
     "execution_count": 172,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_chrom.obs.head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Setup memento with the treatment and covariates\n",
    "\n",
    "Only select the genes defined to be in chromosome 1 above for testing, via the `gene_list` parameter."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ssm-user/anaconda3/envs/single_cell/lib/python3.8/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n",
      "/home/ssm-user/anaconda3/envs/single_cell/lib/python3.8/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  df_sub[k].cat.remove_unused_categories(inplace=True)\n"
     ]
    }
   ],
   "source": [
    "adata_chrom.obs['capture_rate'] = 0.07\n",
    "memento.setup_memento(adata_chrom, q_column='capture_rate')\n",
    "memento.create_groups(adata_chrom, label_columns=['stim', 'chr_expr_avg'])\n",
    "memento.compute_1d_moments(adata_chrom, min_perc_group=.7, gene_list=chr1_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(5341, 117)"
      ]
     },
     "execution_count": 176,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_chrom.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Perform 1D hypothesis testing\n",
    "\n",
    "Sample metadata has all the columns we are interested in, both the treatment and the covariates. \n",
    "\n",
    "We will separate it out into the treatment and covariate DataFrames."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_meta = memento.get_groups(adata_chrom)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The covariate DataFrame - pick the covariate columns\n",
    "cov_df = sample_meta[['chr_expr_avg']]\n",
    "\n",
    "# The treatment DataFrame - pick the treatment column\n",
    "treat_df = sample_meta[['stim']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Parallel(n_jobs=14)]: Using backend LokyBackend with 14 concurrent workers.\n",
      "[Parallel(n_jobs=14)]: Done  22 tasks      | elapsed:    1.8s\n",
      "[Parallel(n_jobs=14)]: Done 117 out of 117 | elapsed:    9.4s finished\n"
     ]
    }
   ],
   "source": [
    "memento.ht_1d_moments(\n",
    "    adata_chrom, \n",
    "    treatment=treat_df,\n",
    "    covariate=cov_df,\n",
    "    resampling='bootstrap',\n",
    "    num_boot=5000, \n",
    "    verbose=1,\n",
    "    num_cpus=14)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "result_1d = memento.get_1d_ht_result(adata_chrom)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>tx</th>\n",
       "      <th>de_coef</th>\n",
       "      <th>de_se</th>\n",
       "      <th>de_pval</th>\n",
       "      <th>dv_coef</th>\n",
       "      <th>dv_se</th>\n",
       "      <th>dv_pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>SAT1</td>\n",
       "      <td>stim</td>\n",
       "      <td>1.315939</td>\n",
       "      <td>0.030836</td>\n",
       "      <td>3.916619e-08</td>\n",
       "      <td>0.910213</td>\n",
       "      <td>0.144714</td>\n",
       "      <td>4.629241e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>PLSCR1</td>\n",
       "      <td>stim</td>\n",
       "      <td>1.491328</td>\n",
       "      <td>0.036718</td>\n",
       "      <td>6.500286e-08</td>\n",
       "      <td>-1.052557</td>\n",
       "      <td>0.227541</td>\n",
       "      <td>1.186501e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>228</th>\n",
       "      <td>APOBEC3A</td>\n",
       "      <td>stim</td>\n",
       "      <td>3.364282</td>\n",
       "      <td>0.063989</td>\n",
       "      <td>3.375917e-07</td>\n",
       "      <td>-2.178124</td>\n",
       "      <td>0.126437</td>\n",
       "      <td>1.232477e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>RNF114</td>\n",
       "      <td>stim</td>\n",
       "      <td>1.344620</td>\n",
       "      <td>0.069991</td>\n",
       "      <td>6.405498e-07</td>\n",
       "      <td>0.740793</td>\n",
       "      <td>0.414568</td>\n",
       "      <td>7.758448e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>186</th>\n",
       "      <td>CCL2</td>\n",
       "      <td>stim</td>\n",
       "      <td>0.953319</td>\n",
       "      <td>0.044157</td>\n",
       "      <td>9.928827e-07</td>\n",
       "      <td>-1.521958</td>\n",
       "      <td>0.060431</td>\n",
       "      <td>1.972572e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>CD47</td>\n",
       "      <td>stim</td>\n",
       "      <td>0.838913</td>\n",
       "      <td>0.053488</td>\n",
       "      <td>1.030917e-06</td>\n",
       "      <td>0.135161</td>\n",
       "      <td>0.279520</td>\n",
       "      <td>6.284743e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>GBP1</td>\n",
       "      <td>stim</td>\n",
       "      <td>1.806011</td>\n",
       "      <td>0.053052</td>\n",
       "      <td>2.032212e-06</td>\n",
       "      <td>-0.616521</td>\n",
       "      <td>0.157682</td>\n",
       "      <td>1.528063e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>HLA-A</td>\n",
       "      <td>stim</td>\n",
       "      <td>0.266386</td>\n",
       "      <td>0.018729</td>\n",
       "      <td>2.067598e-06</td>\n",
       "      <td>-0.024669</td>\n",
       "      <td>0.140585</td>\n",
       "      <td>8.576285e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>TMEM60</td>\n",
       "      <td>stim</td>\n",
       "      <td>1.122225</td>\n",
       "      <td>0.086455</td>\n",
       "      <td>4.376555e-06</td>\n",
       "      <td>0.404528</td>\n",
       "      <td>0.451777</td>\n",
       "      <td>3.931214e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>IFITM3</td>\n",
       "      <td>stim</td>\n",
       "      <td>3.349291</td>\n",
       "      <td>0.053356</td>\n",
       "      <td>5.734026e-06</td>\n",
       "      <td>-3.070060</td>\n",
       "      <td>0.129696</td>\n",
       "      <td>2.348415e-08</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         gene    tx   de_coef     de_se       de_pval   dv_coef     dv_se  \\\n",
       "96       SAT1  stim  1.315939  0.030836  3.916619e-08  0.910213  0.144714   \n",
       "52     PLSCR1  stim  1.491328  0.036718  6.500286e-08 -1.052557  0.227541   \n",
       "228  APOBEC3A  stim  3.364282  0.063989  3.375917e-07 -2.178124  0.126437   \n",
       "200    RNF114  stim  1.344620  0.069991  6.405498e-07  0.740793  0.414568   \n",
       "186      CCL2  stim  0.953319  0.044157  9.928827e-07 -1.521958  0.060431   \n",
       "48       CD47  stim  0.838913  0.053488  1.030917e-06  0.135161  0.279520   \n",
       "8        GBP1  stim  1.806011  0.053052  2.032212e-06 -0.616521  0.157682   \n",
       "74      HLA-A  stim  0.266386  0.018729  2.067598e-06 -0.024669  0.140585   \n",
       "90     TMEM60  stim  1.122225  0.086455  4.376555e-06  0.404528  0.451777   \n",
       "122    IFITM3  stim  3.349291  0.053356  5.734026e-06 -3.070060  0.129696   \n",
       "\n",
       "          dv_pval  \n",
       "96   4.629241e-05  \n",
       "52   1.186501e-03  \n",
       "228  1.232477e-06  \n",
       "200  7.758448e-02  \n",
       "186  1.972572e-06  \n",
       "48   6.284743e-01  \n",
       "8    1.528063e-03  \n",
       "74   8.576285e-01  \n",
       "90   3.931214e-01  \n",
       "122  2.348415e-08  "
      ]
     },
     "execution_count": 183,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_1d.query('de_coef > 0').sort_values('de_pval').head(10)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
