{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Using `schypo` to analyze Interferon-B response in monocytes\n",
    "\n",
    "To install `schypo` in the pre-release version (for Ye Lab members), install it directly from github"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "from schypo import schypo, util"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig_path = '/data/home/Github/scrna-parameter-estimation/figures/fig4/'\n",
    "data_path = '/data/parameter_estimation/'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read IFN data and filter for monocytes\n",
    "\n",
    "For `schypo`, we need the raw count matrix. Preferrably, feed the one with all genes so that we can choose what genes to look at. \n",
    "\n",
    "One of the columns in `adata.obs` should be the discrete groups to compare mean, variability, and co-variability across. In this case, it's called `stim`. \n",
    "\n",
    "The column containing the covariate that you want p-values for should either:\n",
    "- Be binary (aka the column only contains two unique values, such as 'A' and 'B'. Here, the values are either 'stim' or 'ctrl'.\n",
    "- Be numeric (aka the column contains -1, 0, -1 for each genotype value). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AnnData object with n_obs × n_vars = 5341 × 35635 \n",
      "    obs: 'tsne1', 'tsne2', 'ind', 'stim', 'cluster', 'cell', 'multiplets', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'total_counts_hb', 'log1p_total_counts_hb', 'pct_counts_hb', 'cell_type'\n",
      "    var: 'gene_ids', 'mt', 'hb', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts'\n",
      "    uns: 'cell_type_colors'\n",
      "    obsm: 'X_tsne'\n"
     ]
    }
   ],
   "source": [
    "adata = sc.read(data_path + 'interferon_filtered.h5ad')\n",
    "adata = adata[adata.obs.cell == 'CD14+ Monocytes'].copy()\n",
    "print(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ind</th>\n",
       "      <th>stim</th>\n",
       "      <th>cell</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CCTTTAGATGGAAA-1</th>\n",
       "      <td>1039</td>\n",
       "      <td>ctrl</td>\n",
       "      <td>CD14+ Monocytes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGCCATTGCTGGTA-1</th>\n",
       "      <td>1488</td>\n",
       "      <td>stim</td>\n",
       "      <td>CD14+ Monocytes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCGTGCGCCTT-1</th>\n",
       "      <td>1015</td>\n",
       "      <td>ctrl</td>\n",
       "      <td>CD14+ Monocytes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGGAGTGTATTCC-1</th>\n",
       "      <td>1244</td>\n",
       "      <td>ctrl</td>\n",
       "      <td>CD14+ Monocytes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTGTACGACATGAC-1</th>\n",
       "      <td>107</td>\n",
       "      <td>ctrl</td>\n",
       "      <td>CD14+ Monocytes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   ind  stim             cell\n",
       "index                                        \n",
       "CCTTTAGATGGAAA-1  1039  ctrl  CD14+ Monocytes\n",
       "CGCCATTGCTGGTA-1  1488  stim  CD14+ Monocytes\n",
       "AAACCGTGCGCCTT-1  1015  ctrl  CD14+ Monocytes\n",
       "TTGGAGTGTATTCC-1  1244  ctrl  CD14+ Monocytes\n",
       "GTGTACGACATGAC-1   107  ctrl  CD14+ Monocytes"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs[['ind', 'stim', 'cell']].sample(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create groups for hypothesis testing and compute 1D parameters\n",
    "\n",
    "`schypo` creates groups of cells based on anything that should be considered a reasonable group; here, we just divide the cells into `stim` and `ctrl`. But we can easily further divide the cells into individuals by adding the `ind` column to the `label_columns` argument when calling `create_groups`.\n",
    "\n",
    "`q` is the rough estimate of the overall UMI efficiency across both sampling and sequencing. If `s` is the sequencing saturation, multiply `s` by 0.07 for 10X v1, 0.15 for v2, and 0.25 for v3. \n",
    "\n",
    "By default, `schypo` will consider all genes whose expression is high enough to calculate an accurate variance. If you wish to include less genes, increase `filter_mean_thresh`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "schypo.create_groups(adata, label_columns=['stim'], inplace=True, q=0.07)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "schypo.compute_1d_moments(\n",
    "    adata, \n",
    "    inplace=True, \n",
    "    filter_genes=True, # Filter the AnnData object\n",
    "    filter_mean_thresh=0.05, # minimum raw mean of each gene within a group for the gene to be considered \n",
    "    min_perc_group=.9) # percentage of groups that satisfy the condition for a gene to be considered. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Perform 1D hypothesis testing\n",
    "\n",
    "`formula_like` determines the linear model that is used for hypothesis testing, while `cov_column` is used to pick out the variable that you actually want p-values for. \n",
    "\n",
    "`num_cpus` controls how many CPUs to parallelize this operation for. In general, I recommend using 3-6 CPUs for reasonable peformance on any of the AWS machines that we have access to (I'm currently using a c5.2xlarge instance (8 vCPUs). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Parallel(n_jobs=6)]: Using backend LokyBackend with 6 concurrent workers.\n",
      "[Parallel(n_jobs=6)]: Done  38 tasks      | elapsed:    3.1s\n",
      "[Parallel(n_jobs=6)]: Done 188 tasks      | elapsed:   10.2s\n",
      "[Parallel(n_jobs=6)]: Done 438 tasks      | elapsed:   20.9s\n",
      "[Parallel(n_jobs=6)]: Done 788 tasks      | elapsed:   36.5s\n",
      "[Parallel(n_jobs=6)]: Done 1238 tasks      | elapsed:   57.2s\n",
      "[Parallel(n_jobs=6)]: Done 1788 tasks      | elapsed:  1.4min\n",
      "[Parallel(n_jobs=6)]: Done 2337 out of 2337 | elapsed:  1.8min finished\n"
     ]
    }
   ],
   "source": [
    "schypo.ht_1d_moments(\n",
    "    adata, \n",
    "    formula_like='1 + stim',\n",
    "    cov_column='stim', \n",
    "    num_boot=5000, \n",
    "    verbose=1,\n",
    "    num_cpus=6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_1d = schypo.get_1d_ht_result(adata)\n",
    "result_1d['de_fdr'] = util._fdrcorrect(result_1d['de_pval'])\n",
    "result_1d['dv_fdr'] = util._fdrcorrect(result_1d['dv_pval'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>de_coef</th>\n",
       "      <th>de_pval</th>\n",
       "      <th>dv_coef</th>\n",
       "      <th>dv_pval</th>\n",
       "      <th>de_fdr</th>\n",
       "      <th>dv_fdr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>958</th>\n",
       "      <td>LAMTOR4</td>\n",
       "      <td>-0.664377</td>\n",
       "      <td>5.416919e-22</td>\n",
       "      <td>0.162332</td>\n",
       "      <td>0.664267</td>\n",
       "      <td>1.265934e-18</td>\n",
       "      <td>0.871641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>444</th>\n",
       "      <td>BRK1</td>\n",
       "      <td>-1.091400</td>\n",
       "      <td>3.158007e-14</td>\n",
       "      <td>-0.544764</td>\n",
       "      <td>0.175565</td>\n",
       "      <td>3.690131e-11</td>\n",
       "      <td>0.568951</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2196</th>\n",
       "      <td>PLAUR</td>\n",
       "      <td>-1.104445</td>\n",
       "      <td>6.723418e-13</td>\n",
       "      <td>-0.052713</td>\n",
       "      <td>0.807838</td>\n",
       "      <td>5.237542e-10</td>\n",
       "      <td>0.917356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204</th>\n",
       "      <td>CD48</td>\n",
       "      <td>0.718466</td>\n",
       "      <td>3.825761e-12</td>\n",
       "      <td>-0.757255</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>2.235201e-09</td>\n",
       "      <td>0.001010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1384</th>\n",
       "      <td>COX8A</td>\n",
       "      <td>-0.635169</td>\n",
       "      <td>7.446282e-11</td>\n",
       "      <td>0.452424</td>\n",
       "      <td>0.269546</td>\n",
       "      <td>3.480392e-08</td>\n",
       "      <td>0.647409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1403</th>\n",
       "      <td>DRAP1</td>\n",
       "      <td>0.801873</td>\n",
       "      <td>8.985642e-11</td>\n",
       "      <td>-0.292609</td>\n",
       "      <td>0.316737</td>\n",
       "      <td>3.499908e-08</td>\n",
       "      <td>0.697000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>696</th>\n",
       "      <td>UBE2B</td>\n",
       "      <td>0.232309</td>\n",
       "      <td>2.620131e-10</td>\n",
       "      <td>-0.916129</td>\n",
       "      <td>0.353129</td>\n",
       "      <td>8.747494e-08</td>\n",
       "      <td>0.720164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236</th>\n",
       "      <td>RNPEP</td>\n",
       "      <td>-0.765233</td>\n",
       "      <td>6.329750e-10</td>\n",
       "      <td>0.969703</td>\n",
       "      <td>0.027594</td>\n",
       "      <td>1.643625e-07</td>\n",
       "      <td>0.247082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>267</th>\n",
       "      <td>RPS7</td>\n",
       "      <td>-0.994337</td>\n",
       "      <td>6.168798e-10</td>\n",
       "      <td>0.300835</td>\n",
       "      <td>0.149570</td>\n",
       "      <td>1.643625e-07</td>\n",
       "      <td>0.536936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1678</th>\n",
       "      <td>COX16</td>\n",
       "      <td>-0.522020</td>\n",
       "      <td>9.780651e-10</td>\n",
       "      <td>0.293606</td>\n",
       "      <td>0.626675</td>\n",
       "      <td>2.285738e-07</td>\n",
       "      <td>0.858463</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         gene   de_coef       de_pval   dv_coef   dv_pval        de_fdr  \\\n",
       "958   LAMTOR4 -0.664377  5.416919e-22  0.162332  0.664267  1.265934e-18   \n",
       "444      BRK1 -1.091400  3.158007e-14 -0.544764  0.175565  3.690131e-11   \n",
       "2196    PLAUR -1.104445  6.723418e-13 -0.052713  0.807838  5.237542e-10   \n",
       "204      CD48  0.718466  3.825761e-12 -0.757255  0.000006  2.235201e-09   \n",
       "1384    COX8A -0.635169  7.446282e-11  0.452424  0.269546  3.480392e-08   \n",
       "1403    DRAP1  0.801873  8.985642e-11 -0.292609  0.316737  3.499908e-08   \n",
       "696     UBE2B  0.232309  2.620131e-10 -0.916129  0.353129  8.747494e-08   \n",
       "236     RNPEP -0.765233  6.329750e-10  0.969703  0.027594  1.643625e-07   \n",
       "267      RPS7 -0.994337  6.168798e-10  0.300835  0.149570  1.643625e-07   \n",
       "1678    COX16 -0.522020  9.780651e-10  0.293606  0.626675  2.285738e-07   \n",
       "\n",
       "        dv_fdr  \n",
       "958   0.871641  \n",
       "444   0.568951  \n",
       "2196  0.917356  \n",
       "204   0.001010  \n",
       "1384  0.647409  \n",
       "1403  0.697000  \n",
       "696   0.720164  \n",
       "236   0.247082  \n",
       "267   0.536936  \n",
       "1678  0.858463  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_1d.query('de_fdr < 0.1').sort_values('de_fdr').head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>de_coef</th>\n",
       "      <th>de_pval</th>\n",
       "      <th>dv_coef</th>\n",
       "      <th>dv_pval</th>\n",
       "      <th>de_fdr</th>\n",
       "      <th>dv_fdr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1436</th>\n",
       "      <td>CARD16</td>\n",
       "      <td>0.473250</td>\n",
       "      <td>2.476202e-06</td>\n",
       "      <td>-0.917231</td>\n",
       "      <td>1.803329e-09</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.000004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1041</th>\n",
       "      <td>RPS4X</td>\n",
       "      <td>-0.645258</td>\n",
       "      <td>1.046123e-04</td>\n",
       "      <td>0.478713</td>\n",
       "      <td>1.866258e-07</td>\n",
       "      <td>0.000165</td>\n",
       "      <td>0.000218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1100</th>\n",
       "      <td>IDO1</td>\n",
       "      <td>3.859922</td>\n",
       "      <td>1.835622e-08</td>\n",
       "      <td>-1.506381</td>\n",
       "      <td>3.362157e-07</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1905</th>\n",
       "      <td>CCL2</td>\n",
       "      <td>1.450735</td>\n",
       "      <td>3.386626e-07</td>\n",
       "      <td>-1.074444</td>\n",
       "      <td>5.098625e-07</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ISG15</td>\n",
       "      <td>4.524188</td>\n",
       "      <td>3.335918e-07</td>\n",
       "      <td>-2.205038</td>\n",
       "      <td>1.429808e-06</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179</th>\n",
       "      <td>S100A4</td>\n",
       "      <td>-0.625340</td>\n",
       "      <td>4.925540e-05</td>\n",
       "      <td>-0.546267</td>\n",
       "      <td>1.638266e-06</td>\n",
       "      <td>0.000081</td>\n",
       "      <td>0.000638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1317</th>\n",
       "      <td>IFITM3</td>\n",
       "      <td>3.292916</td>\n",
       "      <td>3.296029e-08</td>\n",
       "      <td>-2.206570</td>\n",
       "      <td>2.058045e-06</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1160</th>\n",
       "      <td>RPS6</td>\n",
       "      <td>-0.825058</td>\n",
       "      <td>7.754184e-08</td>\n",
       "      <td>0.549673</td>\n",
       "      <td>3.158067e-06</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>591</th>\n",
       "      <td>IL8</td>\n",
       "      <td>-2.041506</td>\n",
       "      <td>4.786861e-07</td>\n",
       "      <td>0.865848</td>\n",
       "      <td>3.182921e-06</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.000826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>723</th>\n",
       "      <td>RPS14</td>\n",
       "      <td>-0.399055</td>\n",
       "      <td>6.884207e-07</td>\n",
       "      <td>0.419926</td>\n",
       "      <td>3.840718e-06</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.000898</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene   de_coef       de_pval   dv_coef       dv_pval    de_fdr  \\\n",
       "1436  CARD16  0.473250  2.476202e-06 -0.917231  1.803329e-09  0.000009   \n",
       "1041   RPS4X -0.645258  1.046123e-04  0.478713  1.866258e-07  0.000165   \n",
       "1100    IDO1  3.859922  1.835622e-08 -1.506381  3.362157e-07  0.000001   \n",
       "1905    CCL2  1.450735  3.386626e-07 -1.074444  5.098625e-07  0.000003   \n",
       "0      ISG15  4.524188  3.335918e-07 -2.205038  1.429808e-06  0.000003   \n",
       "179   S100A4 -0.625340  4.925540e-05 -0.546267  1.638266e-06  0.000081   \n",
       "1317  IFITM3  3.292916  3.296029e-08 -2.206570  2.058045e-06  0.000001   \n",
       "1160    RPS6 -0.825058  7.754184e-08  0.549673  3.158067e-06  0.000002   \n",
       "591      IL8 -2.041506  4.786861e-07  0.865848  3.182921e-06  0.000004   \n",
       "723    RPS14 -0.399055  6.884207e-07  0.419926  3.840718e-06  0.000005   \n",
       "\n",
       "        dv_fdr  \n",
       "1436  0.000004  \n",
       "1041  0.000218  \n",
       "1100  0.000262  \n",
       "1905  0.000298  \n",
       "0     0.000638  \n",
       "179   0.000638  \n",
       "1317  0.000687  \n",
       "1160  0.000826  \n",
       "591   0.000826  \n",
       "723   0.000898  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_1d.query('dv_fdr < 0.1').sort_values('dv_fdr').head(10)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Perform 2D hypothesis testing\n",
    "\n",
    "For differential co-variability testing, we can specify which genes you want to perform HT on. It takes two lists of genes, `gene_1` and `gene_2`, and it will calculate correlations between every gene in `gene_1` and every gene in `gene_2`. Here, we focus on 3 transcription factors and their correlations to rest of the transcriptome. \n",
    "\n",
    "Similar to the 1D case, 2D hypothesis testing scales with the number of pairs of genes to test. If you have a smaller set of candidate genes, it will run faster."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "schypo.compute_2d_moments(\n",
    "    adata, \n",
    "    ['JUN', 'FOS', 'STAT1'], \n",
    "    adata.var.index.tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Parallel(n_jobs=6)]: Using backend LokyBackend with 6 concurrent workers.\n",
      "[Parallel(n_jobs=6)]: Done  20 tasks      | elapsed:    3.2s\n",
      "[Parallel(n_jobs=6)]: Done 116 tasks      | elapsed:   10.2s\n",
      "[Parallel(n_jobs=6)]: Done 276 tasks      | elapsed:   22.3s\n",
      "[Parallel(n_jobs=6)]: Done 500 tasks      | elapsed:   37.8s\n",
      "[Parallel(n_jobs=6)]: Done 788 tasks      | elapsed:   58.7s\n",
      "[Parallel(n_jobs=6)]: Done 1140 tasks      | elapsed:  1.4min\n",
      "[Parallel(n_jobs=6)]: Done 1556 tasks      | elapsed:  1.9min\n",
      "[Parallel(n_jobs=6)]: Done 2036 tasks      | elapsed:  2.5min\n",
      "[Parallel(n_jobs=6)]: Done 2580 tasks      | elapsed:  3.2min\n",
      "[Parallel(n_jobs=6)]: Done 3188 tasks      | elapsed:  3.9min\n",
      "[Parallel(n_jobs=6)]: Done 3860 tasks      | elapsed:  4.6min\n",
      "[Parallel(n_jobs=6)]: Done 4596 tasks      | elapsed:  5.5min\n",
      "[Parallel(n_jobs=6)]: Done 5396 tasks      | elapsed:  6.5min\n",
      "[Parallel(n_jobs=6)]: Done 6260 tasks      | elapsed:  7.7min\n",
      "[Parallel(n_jobs=6)]: Done 7005 out of 7005 | elapsed:  8.7min finished\n"
     ]
    }
   ],
   "source": [
    "schypo.ht_2d_moments(\n",
    "    adata, \n",
    "    formula_like='1 + stim', \n",
    "    cov_column='stim', \n",
    "    num_cpus=6, \n",
    "    num_boot=5000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_2d = schypo.get_2d_ht_result(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_1</th>\n",
       "      <th>gene_2</th>\n",
       "      <th>corr_coef</th>\n",
       "      <th>corr_pval</th>\n",
       "      <th>corr_fdr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5762</th>\n",
       "      <td>STAT1</td>\n",
       "      <td>DOK2</td>\n",
       "      <td>-0.408536</td>\n",
       "      <td>1.124117e-06</td>\n",
       "      <td>0.003939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3023</th>\n",
       "      <td>FOS</td>\n",
       "      <td>HINT1</td>\n",
       "      <td>-0.322596</td>\n",
       "      <td>6.272982e-07</td>\n",
       "      <td>0.003939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>JUN</td>\n",
       "      <td>PTP4A2</td>\n",
       "      <td>-0.298890</td>\n",
       "      <td>2.131984e-06</td>\n",
       "      <td>0.004980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3862</th>\n",
       "      <td>FOS</td>\n",
       "      <td>GLIPR1</td>\n",
       "      <td>-0.287573</td>\n",
       "      <td>4.682828e-06</td>\n",
       "      <td>0.006978</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3536</th>\n",
       "      <td>FOS</td>\n",
       "      <td>FKBP15</td>\n",
       "      <td>-0.320691</td>\n",
       "      <td>4.978389e-06</td>\n",
       "      <td>0.006978</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1437</th>\n",
       "      <td>JUN</td>\n",
       "      <td>RDX</td>\n",
       "      <td>-0.292184</td>\n",
       "      <td>1.518471e-05</td>\n",
       "      <td>0.016838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4052</th>\n",
       "      <td>FOS</td>\n",
       "      <td>KLF13</td>\n",
       "      <td>-0.384214</td>\n",
       "      <td>1.887668e-05</td>\n",
       "      <td>0.016838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1412</th>\n",
       "      <td>JUN</td>\n",
       "      <td>TCIRG1</td>\n",
       "      <td>0.437781</td>\n",
       "      <td>1.922116e-05</td>\n",
       "      <td>0.016838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6369</th>\n",
       "      <td>STAT1</td>\n",
       "      <td>LGMN</td>\n",
       "      <td>0.264664</td>\n",
       "      <td>2.542207e-05</td>\n",
       "      <td>0.019795</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3551</th>\n",
       "      <td>FOS</td>\n",
       "      <td>ENG</td>\n",
       "      <td>-0.350422</td>\n",
       "      <td>2.383069e-04</td>\n",
       "      <td>0.167005</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene_1  gene_2  corr_coef     corr_pval  corr_fdr\n",
       "5762  STAT1    DOK2  -0.408536  1.124117e-06  0.003939\n",
       "3023    FOS   HINT1  -0.322596  6.272982e-07  0.003939\n",
       "62      JUN  PTP4A2  -0.298890  2.131984e-06  0.004980\n",
       "3862    FOS  GLIPR1  -0.287573  4.682828e-06  0.006978\n",
       "3536    FOS  FKBP15  -0.320691  4.978389e-06  0.006978\n",
       "1437    JUN     RDX  -0.292184  1.518471e-05  0.016838\n",
       "4052    FOS   KLF13  -0.384214  1.887668e-05  0.016838\n",
       "1412    JUN  TCIRG1   0.437781  1.922116e-05  0.016838\n",
       "6369  STAT1    LGMN   0.264664  2.542207e-05  0.019795\n",
       "3551    FOS     ENG  -0.350422  2.383069e-04  0.167005"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_2d.sort_values('corr_fdr').head(10)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
