{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "56a4b02a",
   "metadata": {},
   "source": [
    "### ATAC enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09c7325d",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(data.table)\n",
    "# library(qvalue)\n",
    "library(bedr)\n",
    "options(scipen = 999)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd5614c5",
   "metadata": {},
   "source": [
    "### Copy gracie's attempt with memento  eqtls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f89118e",
   "metadata": {},
   "outputs": [],
   "source": [
    "for (ct in c('T4', 'cM', 'ncM', 'T8', 'B', 'NK')){\n",
    "print(paste('starting....',ct))\n",
    "library(data.table)\n",
    "library(qvalue)\n",
    "library(bedr)\n",
    "options(scipen = 999)\n",
    "# args=commandArgs(TRUE)\n",
    "\n",
    "res_raw=data.frame(fread(paste('/data_volume/memento/lupus/full_analysis/memento/100kb/eur_',ct, '.csv', sep='')))\n",
    "#res_raw=data.frame(fread(args[1]))[-1]\n",
    "#emp=fread(args[2])\n",
    "#delta=fread(args[3])\n",
    "out=paste('/data_volume/memento/lupus/atac_enrichment/100kb/', 'eur_memento_',ct, '.out', sep='')\n",
    "#expr_genes=read.table(args[3],header=F)\n",
    "\n",
    "print('gene')\n",
    "#print(head(expr_genes))\n",
    "#print(dim(expr_genes))\n",
    "#print(head(emp))\n",
    "#emp$FDR=qvalue(emp$Empirical)$qvalues\n",
    "#thresh=max(emp$Pvalue[which(emp$FDR < 0.1)])\n",
    "\n",
    "#run if not corrected by celltype\n",
    "#fdr=qvalue(res_raw$pvalue)\n",
    "#print(fdr)\n",
    "#res_raw$FDR=fdr$qvalues\n",
    "\n",
    "#filter FDR\n",
    "#res=res_raw[res_raw$FDR <= 0.1,]\n",
    "#snps gene statistic pvalue FDR beta qvalue\n",
    "\n",
    "\n",
    "#colnames(res_raw)=c(\"SNP\",\"gene\",\"statistic\",\"p.value\",\"FDR\",\"beta\")\n",
    "#take in metasoft data, create atac enrichments from fixed effects\n",
    "colnames(res_raw)=c('SNP', 'gene', 'statistic', 'p.value', 'FDR', 'beta')\n",
    "# res_raw$gene=sapply(strsplit(res_raw$RSID, '_'), '[', 2)\n",
    "\n",
    "#colnames(res_raw)=c(\"SNP\",\"gene\",\"statistic\",\"p.value\",\"FDR\",\"beta\",\"qvalue\")\n",
    "#res=res_raw\n",
    "#filter for expressed genes\n",
    "# print(head(res_raw))\n",
    "# print(dim(res_raw))\n",
    "#res_expr=res_raw[res_raw$gene %in% expr_genes$V1,]\n",
    "#print(dim((res_expr)))\n",
    "#res_noexpr=res_raw[!(res_raw$gene %in% expr_genes$V1),]\n",
    "\n",
    "# print('expr')\n",
    "#print(dim(res_expr))\n",
    "#print(head(res_expr))\n",
    "# print('no gene')\n",
    "#print(head(res_noexpr))\n",
    "\n",
    "#num_no=length(res_noexpr$gene)\n",
    "#print(num_no)\n",
    "#nullp=runif(num_no, min = 0, max = 1)\n",
    "#nullp=rep(1,num_no)\n",
    "#res_noexpr$p.value=nullp\n",
    "#res=rbind(res_expr,res_noexpr)\n",
    "#print(head(res_expr))\n",
    "#print(dim(res_expr))\n",
    "res=res_raw\n",
    "# print(dim(res))\n",
    "# print(tail(res))\n",
    "#atac=fread('/ye/yelabstore2/10x.lupus/process.scRNAseq/eqtls/S9_lineage_groups.txt.gz')\n",
    "#atac=fread('simple_sorted_atac_lineage_groups2.txt.gz')\n",
    "atac=fread('/data_volume/memento/lupus/atac_enrichment/sorted_simple_atac_lineage_groups3.bed.gz')\n",
    "\n",
    "#names(atac)=c('chr', 'start', 'end', 'group')\n",
    "#atac.bed=data.frame(chr=atac$chr, start=atac$start, end=atac$stop, group=atac$group)\n",
    "atac.groups=names(table(atac$group))\n",
    "#atac.sorted=bedr.sort.region(atac$chr,atac$start,atac$stop)\n",
    "\n",
    "res$chr=paste('chr', sapply(strsplit(res$SNP, ':'), '[', 1), sep='')\n",
    "res$pos=as.numeric(sapply(strsplit(res$SNP, ':'), '[', 2))\n",
    "\n",
    "#res$chr=paste('chr', sapply(strsplit(res$snp, ':'), '[', 1), sep='')\n",
    "#res$pos=as.numeric(sapply(strsplit(res$snp, ':'), '[', 2))\n",
    "res$id=paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')\n",
    "res.bed.sorted=bedr.sort.region(unique(paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')))\n",
    "\n",
    "#res$chr=paste('chr', sapply(strsplit(res$SNP, ':'), '[', 1), sep='')\n",
    "#res$pos=as.numeric(sapply(strsplit(res$SNP, ':'), '[', 2))\n",
    "#res$id=paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')\n",
    "#res.bed.sorted=bedr.sort.region(unique(paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')))\n",
    "\n",
    "all.df=NULL\n",
    "for(g in c('B', 'myeloid', 'nk', 'T')){\n",
    "    print(g)\n",
    "    peaks=atac[which(atac$group == g), ]\n",
    "    peaks=paste(peaks$chr, ':', peaks$start, '-', peaks$stop, sep='')\n",
    "    overlap.sig.snps=bedr(input=list(a=res.bed.sorted, b=peaks), method='intersect')\n",
    "    in_peak=as.numeric(res.bed.sorted %in% overlap.sig.snps)\n",
    "    pvals=res$\"p.value\"[match(res.bed.sorted, res$id)]\n",
    "    #pvals=res$\"pvalue\"[match(res.bed.sorted, res$id)]\n",
    "    #pvals=res$\"p-value\"[match(res.bed.sorted, res$id)]\n",
    "    mw_res=wilcox.test(pvals ~ in_peak, alternative='greater')\n",
    "    df=data.frame(W=mw_res$statistic, pval=mw_res$p.value, group=g)\n",
    "    all.df=rbind(all.df, df)   \n",
    "}\n",
    "\n",
    "write.table(all.df, file=out, row.names=F, col.names=T, quote=F, sep='\\t')\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "39725b2e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'B'</li><li>'EPI'</li><li>'myeloid'</li><li>'nk'</li><li>'open'</li><li>'progenitor'</li><li>'T'</li><li>'thymocyte'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'B'\n",
       "\\item 'EPI'\n",
       "\\item 'myeloid'\n",
       "\\item 'nk'\n",
       "\\item 'open'\n",
       "\\item 'progenitor'\n",
       "\\item 'T'\n",
       "\\item 'thymocyte'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'B'\n",
       "2. 'EPI'\n",
       "3. 'myeloid'\n",
       "4. 'nk'\n",
       "5. 'open'\n",
       "6. 'progenitor'\n",
       "7. 'T'\n",
       "8. 'thymocyte'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"B\"          \"EPI\"        \"myeloid\"    \"nk\"         \"open\"      \n",
       "[6] \"progenitor\" \"T\"          \"thymocyte\" "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "atac.groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "65443886",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>W</th><th scope=col>pval</th><th scope=col>conf</th><th scope=col>estimate</th><th scope=col>group</th></tr>\n",
       "\t<tr><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>4590296557</td><td>0.767464</td><td>-0.006153402</td><td>-0.001645391</td><td>B</td></tr>\n",
       "\t<tr><td>4590296557</td><td>0.767464</td><td>         Inf</td><td>-0.001645391</td><td>B</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 5\n",
       "\\begin{tabular}{lllll}\n",
       " W & pval & conf & estimate & group\\\\\n",
       " <dbl> & <dbl> & <dbl> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\t 4590296557 & 0.767464 & -0.006153402 & -0.001645391 & B\\\\\n",
       "\t 4590296557 & 0.767464 &          Inf & -0.001645391 & B\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 5\n",
       "\n",
       "| W &lt;dbl&gt; | pval &lt;dbl&gt; | conf &lt;dbl&gt; | estimate &lt;dbl&gt; | group &lt;chr&gt; |\n",
       "|---|---|---|---|---|\n",
       "| 4590296557 | 0.767464 | -0.006153402 | -0.001645391 | B |\n",
       "| 4590296557 | 0.767464 |          Inf | -0.001645391 | B |\n",
       "\n"
      ],
      "text/plain": [
       "  W          pval     conf         estimate     group\n",
       "1 4590296557 0.767464 -0.006153402 -0.001645391 B    \n",
       "2 4590296557 0.767464          Inf -0.001645391 B    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "conf.int"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b4539fa6",
   "metadata": {},
   "source": [
    "### Copy gracie's attempt with merged  eqtls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5cb7be91",
   "metadata": {},
   "outputs": [],
   "source": [
    "for (pop in c('asian')){\n",
    "    for (ct in c('T4', 'cM', 'ncM', 'T8', 'B', 'NK')){\n",
    "        print(ct)\n",
    "        library(data.table)\n",
    "        library(qvalue)\n",
    "        library(bedr)\n",
    "        options(scipen = 999)\n",
    "        # args=commandArgs(TRUE)\n",
    "\n",
    "        res_raw=data.frame(fread(paste('/data_volume/memento/lupus/atac_enrichment/merged_eqtls/',pop,'_',ct, '_merged.tsv', sep='')))\n",
    "        #res_raw=data.frame(fread(args[1]))[-1]\n",
    "        #emp=fread(args[2])\n",
    "        #delta=fread(args[3])\n",
    "        out=paste('/data_volume/memento/lupus/atac_enrichment/gracie/', pop, '_',ct, '_merged.out', sep='')\n",
    "        #expr_genes=read.table(args[3],header=F)\n",
    "\n",
    "        print('gene')\n",
    "        #print(head(expr_genes))\n",
    "        #print(dim(expr_genes))\n",
    "        #print(head(emp))\n",
    "        #emp$FDR=qvalue(emp$Empirical)$qvalues\n",
    "        #thresh=max(emp$Pvalue[which(emp$FDR < 0.1)])\n",
    "\n",
    "        #run if not corrected by celltype\n",
    "        #fdr=qvalue(res_raw$pvalue)\n",
    "        #print(fdr)\n",
    "        #res_raw$FDR=fdr$qvalues\n",
    "\n",
    "        #filter FDR\n",
    "        #res=res_raw[res_raw$FDR <= 0.1,]\n",
    "        #snps gene statistic pvalue FDR beta qvalue\n",
    "\n",
    "\n",
    "        #colnames(res_raw)=c(\"SNP\",\"gene\",\"statistic\",\"p.value\",\"FDR\",\"beta\")\n",
    "        #take in metasoft data, create atac enrichments from fixed effects\n",
    "        colnames(res_raw)=c('SNP', 'gene', 'statistic_mem', 'p-value_mem', 'FDR_mem', 'beta_mem', 'beta_meq', 't-stat_meq', 'p-value_meq', 'FDR_meq')\n",
    "        # res_raw$gene=sapply(strsplit(res_raw$RSID, '_'), '[', 2)\n",
    "\n",
    "        #colnames(res_raw)=c(\"SNP\",\"gene\",\"statistic\",\"p.value\",\"FDR\",\"beta\",\"qvalue\")\n",
    "        #res=res_raw\n",
    "        #filter for expressed genes\n",
    "        print(head(res_raw))\n",
    "        print(dim(res_raw))\n",
    "        #res_expr=res_raw[res_raw$gene %in% expr_genes$V1,]\n",
    "        #print(dim((res_expr)))\n",
    "        #res_noexpr=res_raw[!(res_raw$gene %in% expr_genes$V1),]\n",
    "\n",
    "        print('expr')\n",
    "        #print(dim(res_expr))\n",
    "        #print(head(res_expr))\n",
    "        print('no gene')\n",
    "        #print(head(res_noexpr))\n",
    "\n",
    "        #num_no=length(res_noexpr$gene)\n",
    "        #print(num_no)\n",
    "        #nullp=runif(num_no, min = 0, max = 1)\n",
    "        #nullp=rep(1,num_no)\n",
    "        #res_noexpr$p.value=nullp\n",
    "        #res=rbind(res_expr,res_noexpr)\n",
    "        #print(head(res_expr))\n",
    "        #print(dim(res_expr))\n",
    "        res=res_raw\n",
    "        print(dim(res))\n",
    "        print(tail(res))\n",
    "        #atac=fread('/ye/yelabstore2/10x.lupus/process.scRNAseq/eqtls/S9_lineage_groups.txt.gz')\n",
    "        #atac=fread('simple_sorted_atac_lineage_groups2.txt.gz')\n",
    "        atac=fread('/data_volume/memento/lupus/atac_enrichment/sorted_simple_atac_lineage_groups3.bed.gz')\n",
    "\n",
    "        #names(atac)=c('chr', 'start', 'end', 'group')\n",
    "        #atac.bed=data.frame(chr=atac$chr, start=atac$start, end=atac$stop, group=atac$group)\n",
    "        atac.groups=names(table(atac$group))\n",
    "        #atac.sorted=bedr.sort.region(atac$chr,atac$start,atac$stop)\n",
    "\n",
    "        res$chr=paste('chr', sapply(strsplit(res$SNP, ':'), '[', 1), sep='')\n",
    "        res$pos=as.numeric(sapply(strsplit(res$SNP, ':'), '[', 2))\n",
    "\n",
    "        #res$chr=paste('chr', sapply(strsplit(res$snp, ':'), '[', 1), sep='')\n",
    "        #res$pos=as.numeric(sapply(strsplit(res$snp, ':'), '[', 2))\n",
    "        res$id=paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')\n",
    "        res.bed.sorted=bedr.sort.region(unique(paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')))\n",
    "\n",
    "        #res$chr=paste('chr', sapply(strsplit(res$SNP, ':'), '[', 1), sep='')\n",
    "        #res$pos=as.numeric(sapply(strsplit(res$SNP, ':'), '[', 2))\n",
    "        #res$id=paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')\n",
    "        #res.bed.sorted=bedr.sort.region(unique(paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')))\n",
    "\n",
    "        all.df=NULL\n",
    "        for(g in atac.groups){\n",
    "            print(g)\n",
    "            peaks=atac[which(atac$group == g), ]\n",
    "            peaks=paste(peaks$chr, ':', peaks$start, '-', peaks$stop, sep='')\n",
    "            overlap.sig.snps=bedr(input=list(a=res.bed.sorted, b=peaks), method='intersect')\n",
    "            in_peak=as.numeric(res.bed.sorted %in% overlap.sig.snps)\n",
    "            pvals_mem=res$\"p-value_mem\"[match(res.bed.sorted, res$id)]\n",
    "            pvals_meq=res$\"p-value_meq\"[match(res.bed.sorted, res$id)]\n",
    "            #pvals=res$\"pvalue\"[match(res.bed.sorted, res$id)]\n",
    "            #pvals=res$\"p-value\"[match(res.bed.sorted, res$id)]\n",
    "            mw_res_mem=wilcox.test(pvals_mem ~ in_peak, alternative='greater')\n",
    "            mw_res_meq=wilcox.test(pvals_meq ~ in_peak, alternative='greater')\n",
    "\n",
    "            df=data.frame(W_mem=mw_res_mem$statistic, W_meq=mw_res_meq$statistic, pval_mem=mw_res_mem$p.value, pval_meq=mw_res_meq$p.value,group=g)\n",
    "            all.df=rbind(all.df, df)   \n",
    "        }\n",
    "\n",
    "        write.table(all.df, file=out, row.names=F, col.names=T, quote=F, sep='\\t')\n",
    "\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a570dfdd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "6adad925",
   "metadata": {},
   "source": [
    "### Copy gracie's attempt with mateqtl  eqtls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ca91a7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "for (ct in c('T4', 'cM', 'ncM', 'T8', 'B', 'NK')){\n",
    "print(ct)\n",
    "library(data.table)\n",
    "library(qvalue)\n",
    "library(bedr)\n",
    "options(scipen = 999)\n",
    "# args=commandArgs(TRUE)\n",
    "\n",
    "res_raw=data.frame(fread(paste('/data_volume/memento/lupus/full_analysis/mateqtl/outputs/asian_',ct, '_all_hg19.csv', sep='')))\n",
    "#res_raw=data.frame(fread(args[1]))[-1]\n",
    "#emp=fread(args[2])\n",
    "#delta=fread(args[3])\n",
    "out=paste('/data_volume/memento/lupus/atac_enrichment/gracie/', 'asian_matqetl_',ct, '.out', sep='')\n",
    "#expr_genes=read.table(args[3],header=F)\n",
    "\n",
    "print('gene')\n",
    "#print(head(expr_genes))\n",
    "#print(dim(expr_genes))\n",
    "#print(head(emp))\n",
    "#emp$FDR=qvalue(emp$Empirical)$qvalues\n",
    "#thresh=max(emp$Pvalue[which(emp$FDR < 0.1)])\n",
    "\n",
    "#run if not corrected by celltype\n",
    "#fdr=qvalue(res_raw$pvalue)\n",
    "#print(fdr)\n",
    "#res_raw$FDR=fdr$qvalues\n",
    "\n",
    "#filter FDR\n",
    "#res=res_raw[res_raw$FDR <= 0.1,]\n",
    "#snps gene statistic pvalue FDR beta qvalue\n",
    "\n",
    "\n",
    "#colnames(res_raw)=c(\"SNP\",\"gene\",\"statistic\",\"p.value\",\"FDR\",\"beta\")\n",
    "#take in metasoft data, create atac enrichments from fixed effects\n",
    "colnames(res_raw)=c('SNP', 'gene', 'beta','statistic', 'p.value', 'FDR')\n",
    "# res_raw$gene=sapply(strsplit(res_raw$RSID, '_'), '[', 2)\n",
    "\n",
    "#colnames(res_raw)=c(\"SNP\",\"gene\",\"statistic\",\"p.value\",\"FDR\",\"beta\",\"qvalue\")\n",
    "#res=res_raw\n",
    "#filter for expressed genes\n",
    "print(head(res_raw))\n",
    "print(dim(res_raw))\n",
    "#res_expr=res_raw[res_raw$gene %in% expr_genes$V1,]\n",
    "#print(dim((res_expr)))\n",
    "#res_noexpr=res_raw[!(res_raw$gene %in% expr_genes$V1),]\n",
    "\n",
    "print('expr')\n",
    "#print(dim(res_expr))\n",
    "#print(head(res_expr))\n",
    "print('no gene')\n",
    "#print(head(res_noexpr))\n",
    "\n",
    "#num_no=length(res_noexpr$gene)\n",
    "#print(num_no)\n",
    "#nullp=runif(num_no, min = 0, max = 1)\n",
    "#nullp=rep(1,num_no)\n",
    "#res_noexpr$p.value=nullp\n",
    "#res=rbind(res_expr,res_noexpr)\n",
    "#print(head(res_expr))\n",
    "#print(dim(res_expr))\n",
    "res=res_raw\n",
    "print(dim(res))\n",
    "print(tail(res))\n",
    "#atac=fread('/ye/yelabstore2/10x.lupus/process.scRNAseq/eqtls/S9_lineage_groups.txt.gz')\n",
    "#atac=fread('simple_sorted_atac_lineage_groups2.txt.gz')\n",
    "atac=fread('/data_volume/memento/lupus/atac_enrichment/sorted_simple_atac_lineage_groups3.bed.gz')\n",
    "\n",
    "#names(atac)=c('chr', 'start', 'end', 'group')\n",
    "#atac.bed=data.frame(chr=atac$chr, start=atac$start, end=atac$stop, group=atac$group)\n",
    "atac.groups=names(table(atac$group))\n",
    "#atac.sorted=bedr.sort.region(atac$chr,atac$start,atac$stop)\n",
    "\n",
    "res$chr=paste('chr', sapply(strsplit(res$SNP, ':'), '[', 1), sep='')\n",
    "res$pos=as.numeric(sapply(strsplit(res$SNP, ':'), '[', 2))\n",
    "\n",
    "#res$chr=paste('chr', sapply(strsplit(res$snp, ':'), '[', 1), sep='')\n",
    "#res$pos=as.numeric(sapply(strsplit(res$snp, ':'), '[', 2))\n",
    "res$id=paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')\n",
    "res.bed.sorted=bedr.sort.region(unique(paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')))\n",
    "\n",
    "#res$chr=paste('chr', sapply(strsplit(res$SNP, ':'), '[', 1), sep='')\n",
    "#res$pos=as.numeric(sapply(strsplit(res$SNP, ':'), '[', 2))\n",
    "#res$id=paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')\n",
    "#res.bed.sorted=bedr.sort.region(unique(paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')))\n",
    "\n",
    "all.df=NULL\n",
    "for(g in atac.groups){\n",
    "    print(g)\n",
    "    peaks=atac[which(atac$group == g), ]\n",
    "    peaks=paste(peaks$chr, ':', peaks$start, '-', peaks$stop, sep='')\n",
    "    overlap.sig.snps=bedr(input=list(a=res.bed.sorted, b=peaks), method='intersect')\n",
    "    in_peak=as.numeric(res.bed.sorted %in% overlap.sig.snps)\n",
    "    pvals=res$\"p.value\"[match(res.bed.sorted, res$id)]\n",
    "    #pvals=res$\"pvalue\"[match(res.bed.sorted, res$id)]\n",
    "    #pvals=res$\"p-value\"[match(res.bed.sorted, res$id)]\n",
    "    mw_res=wilcox.test(pvals ~ in_peak, alternative='greater')\n",
    "    df=data.frame(W=mw_res$statistic, pval=mw_res$p.value, group=g)\n",
    "    all.df=rbind(all.df, df)   \n",
    "}\n",
    "\n",
    "write.table(all.df, file=out, row.names=F, col.names=T, quote=F, sep='\\t')\n",
    "\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20466aa9",
   "metadata": {},
   "source": [
    "### Get confidence interval"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "f27f13d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "matches <- c(\n",
    "    'T4'='T',\n",
    "    'cM'='myeloid',\n",
    "    'ncM'='myeloid',\n",
    "    'B'='B',\n",
    "    'NK'='nk'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8b17042",
   "metadata": {},
   "outputs": [],
   "source": [
    "all.df=NULL\n",
    "pop <- 'asian'\n",
    "for (ct in c('T4', 'cM', 'ncM', 'T8', 'B', 'NK')){\n",
    "    print(ct)\n",
    "    library(data.table)\n",
    "    library(qvalue)\n",
    "    library(bedr)\n",
    "    options(scipen = 999)\n",
    "    # args=commandArgs(TRUE)\n",
    "\n",
    "    res_raw=data.frame(fread(paste('/data_volume/memento/lupus/full_analysis/mateqtl/outputs/', pop, '_',ct, '_all_hg19.csv', sep='')))\n",
    "    #res_raw=data.frame(fread(args[1]))[-1]\n",
    "    #emp=fread(args[2])\n",
    "    #delta=fread(args[3])\n",
    "    #expr_genes=read.table(args[3],header=F)\n",
    "\n",
    "    print('gene')\n",
    "    #print(head(expr_genes))\n",
    "    #print(dim(expr_genes))\n",
    "    #print(head(emp))\n",
    "    #emp$FDR=qvalue(emp$Empirical)$qvalues\n",
    "    #thresh=max(emp$Pvalue[which(emp$FDR < 0.1)])\n",
    "\n",
    "    #run if not corrected by celltype\n",
    "    #fdr=qvalue(res_raw$pvalue)\n",
    "    #print(fdr)\n",
    "    #res_raw$FDR=fdr$qvalues\n",
    "\n",
    "    #filter FDR\n",
    "    #res=res_raw[res_raw$FDR <= 0.1,]\n",
    "    #snps gene statistic pvalue FDR beta qvalue\n",
    "\n",
    "\n",
    "    #colnames(res_raw)=c(\"SNP\",\"gene\",\"statistic\",\"p.value\",\"FDR\",\"beta\")\n",
    "    #take in metasoft data, create atac enrichments from fixed effects\n",
    "    colnames(res_raw)=c('SNP', 'gene', 'beta','statistic', 'p.value', 'FDR')\n",
    "    # res_raw$gene=sapply(strsplit(res_raw$RSID, '_'), '[', 2)\n",
    "\n",
    "    #colnames(res_raw)=c(\"SNP\",\"gene\",\"statistic\",\"p.value\",\"FDR\",\"beta\",\"qvalue\")\n",
    "    #res=res_raw\n",
    "    #filter for expressed genes\n",
    "    print(head(res_raw))\n",
    "    print(dim(res_raw))\n",
    "    #res_expr=res_raw[res_raw$gene %in% expr_genes$V1,]\n",
    "    #print(dim((res_expr)))\n",
    "    #res_noexpr=res_raw[!(res_raw$gene %in% expr_genes$V1),]\n",
    "\n",
    "    print('expr')\n",
    "    #print(dim(res_expr))\n",
    "    #print(head(res_expr))\n",
    "    print('no gene')\n",
    "    #print(head(res_noexpr))\n",
    "\n",
    "    #num_no=length(res_noexpr$gene)\n",
    "    #print(num_no)\n",
    "    #nullp=runif(num_no, min = 0, max = 1)\n",
    "    #nullp=rep(1,num_no)\n",
    "    #res_noexpr$p.value=nullp\n",
    "    #res=rbind(res_expr,res_noexpr)\n",
    "    #print(head(res_expr))\n",
    "    #print(dim(res_expr))\n",
    "    res=res_raw\n",
    "    print(dim(res))\n",
    "    print(tail(res))\n",
    "    #atac=fread('/ye/yelabstore2/10x.lupus/process.scRNAseq/eqtls/S9_lineage_groups.txt.gz')\n",
    "    #atac=fread('simple_sorted_atac_lineage_groups2.txt.gz')\n",
    "    atac=fread('/data_volume/memento/lupus/atac_enrichment/sorted_simple_atac_lineage_groups3.bed.gz')\n",
    "\n",
    "    #names(atac)=c('chr', 'start', 'end', 'group')\n",
    "    #atac.bed=data.frame(chr=atac$chr, start=atac$start, end=atac$stop, group=atac$group)\n",
    "    atac.groups=names(table(atac$group))\n",
    "    #atac.sorted=bedr.sort.region(atac$chr,atac$start,atac$stop)\n",
    "\n",
    "    res$chr=paste('chr', sapply(strsplit(res$SNP, ':'), '[', 1), sep='')\n",
    "    res$pos=as.numeric(sapply(strsplit(res$SNP, ':'), '[', 2))\n",
    "\n",
    "    #res$chr=paste('chr', sapply(strsplit(res$snp, ':'), '[', 1), sep='')\n",
    "    #res$pos=as.numeric(sapply(strsplit(res$snp, ':'), '[', 2))\n",
    "    res$id=paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')\n",
    "    res.bed.sorted=bedr.sort.region(unique(paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')))\n",
    "\n",
    "    #res$chr=paste('chr', sapply(strsplit(res$SNP, ':'), '[', 1), sep='')\n",
    "    #res$pos=as.numeric(sapply(strsplit(res$SNP, ':'), '[', 2))\n",
    "    #res$id=paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')\n",
    "    #res.bed.sorted=bedr.sort.region(unique(paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')))\n",
    "    \n",
    "    if (ct == 'T8'){\n",
    "        groups <- c('T', 'nk')\n",
    "    }\n",
    "    else {\n",
    "        groups <- c(matches[ct])\n",
    "    }\n",
    "\n",
    "    for(g in groups){\n",
    "        print(g)\n",
    "        peaks=atac[which(atac$group == g), ]\n",
    "        peaks=paste(peaks$chr, ':', peaks$start, '-', peaks$stop, sep='')\n",
    "        overlap.sig.snps=bedr(input=list(a=res.bed.sorted, b=peaks), method='intersect')\n",
    "        in_peak=as.numeric(res.bed.sorted %in% overlap.sig.snps)\n",
    "        pvals=res$\"p.value\"[match(res.bed.sorted, res$id)]\n",
    "        \n",
    "        write.table(\n",
    "            cbind(in_peak, pvals),\n",
    "            file=paste('/data_volume/memento/lupus/atac_enrichment/matched/mateqtl/', pop, '_',ct, '_', g,'.txt', sep=''), \n",
    "            row.names=F, col.names=T, quote=F, sep='\\t')\n",
    "\n",
    "        #pvals=res$\"pvalue\"[match(res.bed.sorted, res$id)]\n",
    "        #pvals=res$\"p-value\"[match(res.bed.sorted, res$id)]\n",
    "#         mw_res=wilcox.test(pvals ~ in_peak, alternative='two.sided', conf.int=TRUE, conf.level=0.95)\n",
    "#         df=data.frame(W=mw_res$statistic, pval=mw_res$p.value, group=g, ct=ct, lower=mw_res$conf.int[1], upper=mw_res$conf.int[2], estimate=mw_res$estimate)\n",
    "#         all.df=rbind(all.df, df)\n",
    "        \n",
    "    }\n",
    "    \n",
    "\n",
    "}\n",
    "# write.table(all.df, file='asian_mateqtl_confint.txt', row.names=F, col.names=T, quote=F, sep='\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "68d5716e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'ncM'"
      ],
      "text/latex": [
       "'ncM'"
      ],
      "text/markdown": [
       "'ncM'"
      ],
      "text/plain": [
       "[1] \"ncM\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ct"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "ca9c8231",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>0.816888734039885</li><li>0.357804875544383</li><li>0.129175181426451</li><li>0.212469818839699</li><li>0.130747506569229</li><li>0.24333486723542</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 0.816888734039885\n",
       "\\item 0.357804875544383\n",
       "\\item 0.129175181426451\n",
       "\\item 0.212469818839699\n",
       "\\item 0.130747506569229\n",
       "\\item 0.24333486723542\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 0.816888734039885\n",
       "2. 0.357804875544383\n",
       "3. 0.129175181426451\n",
       "4. 0.212469818839699\n",
       "5. 0.130747506569229\n",
       "6. 0.24333486723542\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 0.8168887 0.3578049 0.1291752 0.2124698 0.1307475 0.2433349"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(pvals)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "0de5e389",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>0</li><li>0</li><li>1</li><li>1</li><li>0</li><li>0</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 0\n",
       "\\item 0\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 0\n",
       "\\item 0\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 0\n",
       "2. 0\n",
       "3. 1\n",
       "4. 1\n",
       "5. 0\n",
       "6. 0\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 0 0 1 1 0 0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(in_peak)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "237339e6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 6 × 2 of type dbl</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>pvals</th><th scope=col>in_peak</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>0.8168887</td><td>0</td></tr>\n",
       "\t<tr><td>0.3578049</td><td>0</td></tr>\n",
       "\t<tr><td>0.1291752</td><td>1</td></tr>\n",
       "\t<tr><td>0.2124698</td><td>1</td></tr>\n",
       "\t<tr><td>0.1307475</td><td>0</td></tr>\n",
       "\t<tr><td>0.2433349</td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 6 × 2 of type dbl\n",
       "\\begin{tabular}{ll}\n",
       " pvals & in\\_peak\\\\\n",
       "\\hline\n",
       "\t 0.8168887 & 0\\\\\n",
       "\t 0.3578049 & 0\\\\\n",
       "\t 0.1291752 & 1\\\\\n",
       "\t 0.2124698 & 1\\\\\n",
       "\t 0.1307475 & 0\\\\\n",
       "\t 0.2433349 & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 6 × 2 of type dbl\n",
       "\n",
       "| pvals | in_peak |\n",
       "|---|---|\n",
       "| 0.8168887 | 0 |\n",
       "| 0.3578049 | 0 |\n",
       "| 0.1291752 | 1 |\n",
       "| 0.2124698 | 1 |\n",
       "| 0.1307475 | 0 |\n",
       "| 0.2433349 | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "     pvals     in_peak\n",
       "[1,] 0.8168887 0      \n",
       "[2,] 0.3578049 0      \n",
       "[3,] 0.1291752 1      \n",
       "[4,] 0.2124698 1      \n",
       "[5,] 0.1307475 0      \n",
       "[6,] 0.2433349 0      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(cbind(pvals, in_peak))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "3d2ef9de",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 7</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>W</th><th scope=col>pval</th><th scope=col>group</th><th scope=col>ct</th><th scope=col>lower</th><th scope=col>upper</th><th scope=col>estimate</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>W</th><td>6849773031</td><td>0.09481837</td><td>T      </td><td>T4</td><td>-0.007591921</td><td>0.0005594452</td><td>-0.003388853</td></tr>\n",
       "\t<tr><th scope=row>W1</th><td>4229275845</td><td>0.58932818</td><td>myeloid</td><td>cM</td><td>-0.003186105</td><td>0.0058257938</td><td> 0.001042079</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 7\n",
       "\\begin{tabular}{r|lllllll}\n",
       "  & W & pval & group & ct & lower & upper & estimate\\\\\n",
       "  & <dbl> & <dbl> & <chr> & <chr> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\tW & 6849773031 & 0.09481837 & T       & T4 & -0.007591921 & 0.0005594452 & -0.003388853\\\\\n",
       "\tW1 & 4229275845 & 0.58932818 & myeloid & cM & -0.003186105 & 0.0058257938 &  0.001042079\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 7\n",
       "\n",
       "| <!--/--> | W &lt;dbl&gt; | pval &lt;dbl&gt; | group &lt;chr&gt; | ct &lt;chr&gt; | lower &lt;dbl&gt; | upper &lt;dbl&gt; | estimate &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| W | 6849773031 | 0.09481837 | T       | T4 | -0.007591921 | 0.0005594452 | -0.003388853 |\n",
       "| W1 | 4229275845 | 0.58932818 | myeloid | cM | -0.003186105 | 0.0058257938 |  0.001042079 |\n",
       "\n"
      ],
      "text/plain": [
       "   W          pval       group   ct lower        upper        estimate    \n",
       "W  6849773031 0.09481837 T       T4 -0.007591921 0.0005594452 -0.003388853\n",
       "W1 4229275845 0.58932818 myeloid cM -0.003186105 0.0058257938  0.001042079"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "all.df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26097709",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
